File: iOS/Crisis Connect/Features/Flashlight/FlashlightPattern.swift
import Foundation

struct FlashlightPulse: Equatable {
    let isOn: Bool
    let durationNanoseconds: UInt64
}

enum FlashlightPatterns {
    private static let morseUnit: UInt64 = 200_000_000

    static let sos: [FlashlightPulse] = {
        var pulses: [FlashlightPulse] = []
        func signal(onUnits: UInt64, offUnits: UInt64) {
            pulses.append(.init(isOn: true, durationNanoseconds: onUnits * morseUnit))
            pulses.append(.init(isOn: false, durationNanoseconds: offUnits * morseUnit))
        }

        // SOS is transmitted as one continuous procedural signal: ...---...
        signal(onUnits: 1, offUnits: 1)
        signal(onUnits: 1, offUnits: 1)
        signal(onUnits: 1, offUnits: 1)
        signal(onUnits: 3, offUnits: 1)
        signal(onUnits: 3, offUnits: 1)
        signal(onUnits: 3, offUnits: 1)
        signal(onUnits: 1, offUnits: 1)
        signal(onUnits: 1, offUnits: 1)
        signal(onUnits: 1, offUnits: 7)
        return pulses
    }()

    /// Six visible signals in one minute, followed by a one-minute pause.
    static let emergencyBeacon: [FlashlightPulse] = {
        var pulses: [FlashlightPulse] = []
        for _ in 0..<6 {
            pulses.append(.init(isOn: true, durationNanoseconds: 1_000_000_000))
            pulses.append(.init(isOn: false, durationNanoseconds: 9_000_000_000))
        }
        pulses.append(.init(isOn: false, durationNanoseconds: 60_000_000_000))
        return pulses
    }()

    static func strobe(flashesPerSecond: Int) -> [FlashlightPulse] {
        let safeRate = UInt64(min(max(flashesPerSecond, 1), 3))
        // Round up so timer quantization never pushes the effective rate above 3 Hz.
        let halfPeriod = (500_000_000 + safeRate - 1) / safeRate
        return [
            .init(isOn: true, durationNanoseconds: halfPeriod),
            .init(isOn: false, durationNanoseconds: halfPeriod),
        ]
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/screens/Tools/FlashlightPattern.kt
package com.auralis.crisisconnect.screens.Tools

internal data class FlashlightPulse(
    val isOn: Boolean,
    val durationMillis: Long
)

/** Pure, testable light patterns shared by the flashlight runtime. */
internal object FlashlightPatterns {
    private const val MORSE_UNIT_MILLIS = 200L

    val sos: List<FlashlightPulse> = buildList {
        fun signal(onUnits: Int, offUnits: Int) {
            add(FlashlightPulse(isOn = true, durationMillis = onUnits * MORSE_UNIT_MILLIS))
            add(FlashlightPulse(isOn = false, durationMillis = offUnits * MORSE_UNIT_MILLIS))
        }

        // SOS is transmitted as one continuous procedural signal: ...---...
        signal(1, 1)
        signal(1, 1)
        signal(1, 1)
        signal(3, 1)
        signal(3, 1)
        signal(3, 1)
        signal(1, 1)
        signal(1, 1)
        signal(1, 7)
    }

    /**
     * International mountain distress signal: six visible signals in one minute, followed by a
     * one-minute pause. Each signal starts ten seconds after the previous one.
     */
    val emergencyBeacon: List<FlashlightPulse> = buildList {
        repeat(6) {
            add(FlashlightPulse(isOn = true, durationMillis = 1_000L))
            add(FlashlightPulse(isOn = false, durationMillis = 9_000L))
        }
        add(FlashlightPulse(isOn = false, durationMillis = 60_000L))
    }

    fun strobe(flashesPerSecond: Int): List<FlashlightPulse> {
        val safeRate = flashesPerSecond.coerceIn(1, 3)
        // Round up so timer quantization can never push the effective rate above 3 Hz.
        val halfPeriodMillis = (500L + safeRate - 1) / safeRate
        return listOf(
            FlashlightPulse(isOn = true, durationMillis = halfPeriodMillis),
            FlashlightPulse(isOn = false, durationMillis = halfPeriodMillis)
        )
    }
}
